aten/src/ATen/native/EmbeddingBag.cpp

```diff
@@ -0,0 +1,243 @@
+#include "ATen/ATen.h"
+#include "ATen/Check.h"
+#include "ATen/NativeFunctions.h"
+
+#include <cstring>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <vector>
+
+#include <TH/THBlas.h>
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+namespace at {
+namespace native {
+
+static void make_offset2bag(const Tensor &offsets, const Tensor &indices,
+                            Tensor &offset2bag) {
+  offset2bag.index_fill_(0, offsets, 1);    // offset2bag = [1 0 1 0 1]
+  offset2bag[0] = 0;                        // offset2bag = [0 0 1 0 1]
+  offset2bag = offset2bag.cumsum(0);        //  # offset2bag = [0 0 1 1 2]
+}
+
+static void make_bag_size(const Tensor &offsets, const Tensor &indices,
+                          const int64_t mode, Tensor &bag_size) {
+  if (mode == 1) { // MODE_MEAN
+    if (offsets.sizes()[0] != 1) {
+      bag_size.slice(0, 0, bag_size.sizes()[0] - 1, 1) =
+          offsets.slice(0, 1, offsets.sizes()[0], 1) -
+          offsets.slice(0, 0, offsets.sizes()[0] - 1, 1);
+      bag_size[-1] = indices.sizes()[0] - offsets[-1];
+    }
+  }
+}
+
+static Tensor apply_bag_size(const Tensor &offsets, const Tensor &indices,
+                             const int64_t mode, Tensor &output,
+                             const Tensor &bag_size) {
+  if (mode == 1) { // MODE_MEAN
+    if (offsets.sizes()[0] == 1) {
+      auto bag_size_ = indices.sizes()[0];
+      output /= bag_size_;
+    } else {
+      auto bag_size_ =
+          bag_size.toType(output.type()).unsqueeze(1).expand_as(output);
+      output /= bag_size_;
+    }
+  }
+  return output;
+}
+
+static Tensor apply_bag_size_backward(const Tensor &offsets,
+                                      const Tensor &indices, const int64_t mode,
+                                      Tensor &output, const Tensor &offset2bag,
+                                      const Tensor &bag_size) {
+  if (mode == 1) { // MODE_MEAN
+    if (offsets.sizes()[0] == 1) {
+      auto bag_size = indices.sizes()[0];
+      output /= bag_size;
+    } else {
+      auto bag_size_ = bag_size.toType(output.type())
+                           .unsqueeze(1)
+                           .index_select(0, offset2bag);
+      output /= bag_size_;
+    }
+  }
+  return output;
+}
+
+std::tuple<Tensor, Tensor, Tensor>
+embedding_bag_cpu(const Tensor &weight, const Tensor &indices__,
+                  const Tensor &offsets__, const Tensor &offset2bag__,
+                  const Tensor &bag_size_, const bool scale_grad_by_freq,
```

Comment:
Why is `bag_size` an input? It looks like it's an output computed by `embedding_bag_cpu`.